File: Methods/Java/Methods/Matrix/Gauss.java
package Methods.Matrix;

import Errors.MatrixWithColumnZero;
import Errors.NoDimensionMatrix;
import Errors.NotSquareMatrix;
import Share.Determinant;
import Share.MatrixUtil;

public class Gauss {

    public double[] execute(double[][] matrix ,double[] b, boolean print) throws MatrixWithColumnZero, NotSquareMatrix, NoDimensionMatrix {
        Determinant.determinant(matrix);
        if (print) return gauss(matrix, b);
        return gaussNoPrint(matrix, b);
    }

    private double[] gauss(double[][] matrix ,double[] b) throws MatrixWithColumnZero {

        MatrixUtil.printAmplifiedMatrix(matrix, b);

        for (int i = 0; i < matrix.length - 1; ++i) {

            System.out.println("\u001B[31m" + "Phase " + (i+1) + "\u001B[0m");

            pivot(matrix, i, b);

            MatrixUtil.printAmplifiedMatrix(matrix, b);

            for (int j = i+1; j < matrix.length; ++j) {

                double multiplicand = matrix[j][i] / matrix[i][i];
                elimination(i, j, multiplicand, matrix, b);
            }
            MatrixUtil.printAmplifiedMatrix(matrix, b);
        }

        MatrixUtil.printAmplifiedMatrix(matrix,b);
        return linealRegression(matrix, b);
    }

    private double[] gaussNoPrint(double[][] matrix ,double[] b) throws MatrixWithColumnZero {
        for (int i = 0; i < matrix.length - 1; ++i) {
            pivot(matrix, i, b);
            for (int j = i+1; j < matrix.length; ++j) {
                double multiplicand = matrix[j][i] / matrix[i][i];
                elimination(i, j, multiplicand, matrix, b);
            }
        }
        return linealRegression(matrix, b);
    }

    void pivot(double[][] matrix, int index, double[] b) throws MatrixWithColumnZero {
        if (matrix[index][index] == 0) {
            int i = index + 1;
            while (matrix[i][index] == 0) {
                ++i;
                if (i >= matrix.length) throw new MatrixWithColumnZero("The matrix has a column full of 0");
            }
            MatrixUtil.swapRows(matrix, i, index);
            MatrixUtil.swapValues(b, i, index);
        }
    }

    void elimination(int row1, int row2, double multiplicand, double[][] matrix, double[] b) {
        for (int i = 0; i < matrix[row2].length; ++i) {
            matrix[row2][i] -= multiplicand * matrix[row1][i];
        }
        b[row2] -= multiplicand*b[row1];
    }

    double[] linealRegression(double[][] matrix, double[] b){
        double[] results = new double[matrix.length];
        int term = matrix.length -1;

        for (int i = matrix.length-1; i >= 0; --i) {
            double[] left = new double[matrix.length-i];

            for (int j = 0; j < left.length; j++) {
                left[j] = matrix[i][term-j];
            }
            results[i] = backwardSolve(left, results, b[i]);

        }
        return results;
    }

    double backwardSolve(double[] left, double[] xValues, double right) {
        if (left.length == 1) {
            return right/left[0];
        }else {
            double newRight = right - left[0]*xValues[xValues.length-1];
            return backwardSolve(MatrixUtil.frontCut(left,1), MatrixUtil.cut(xValues,1), newRight);
        }
    }

    @Override
    public String toString() {
        return "Gauss: ";
    }
}

File: Methods/Python/Matrix/Gauss.py
import MatrixUtils as mu
from copy import copy

def gauss(matrix, b):
    for i in range(len(matrix)):
        pivot(matrix, i, b)

        for j in range(i+1,len(matrix)):
            multiplicand = matrix[j][i] / matrix[i][i]
            elimination(i, j, multiplicand, matrix, b)
    return backwardSubstitution(matrix, b)

def pivot(matrix, index, b):
    if matrix[index][index] == 0:
        i = index + 1
        while matrix[i][index] == 0:
            i += 1
        mu.swapRows(matrix, i, index)
        mu.swapValues(b, i, index)
    
    
def elimination(row1, row2, multiplicand, matrix, b):
    for i in range(len(matrix[row2])):
        matrix[row2][i] -= multiplicand*matrix[row1][i]

    b[row2] -= multiplicand*b[row1]


def backwardSubstitution(matrix, b):
    results = [0] * len(matrix)
    term = len(matrix) - 1

    for i in range(len(matrix)-1, -1, -1):
        left = [0] * (len(matrix) - i)
        for j in range(len(left)):
            left[j] = matrix[i][term-j]
        results[i] = backwardSolve(left=left, xValues=results, right=b[i])
    return results


def backwardSolve(left, xValues, right):
    if (len(left) == 1):
        return right/left[0]
    else:
        newRight = right - left[0]*xValues[len(xValues)-1]
        newLeft = copy(left)
        newLeft.pop(0)
        newXValues = copy(xValues)
        newXValues.pop(len(xValues)-1)
        return backwardSolve(left=newLeft, xValues=newXValues, right=newRight)

